In [128]:
import pandas as pd
df = pd.read_csv('/content/tayara (1).csv')

##Investigate duplicate rows than drop them


In [129]:
# Identifying duplicate rows
duplicates = df[df.duplicated()]

print("Duplicate Rows:")
print(len(duplicates.index))

Duplicate Rows:
134


In [130]:
#drop duplicate rows
df = df.drop_duplicates()
display(df)

price transaction                                       title  \
0    490000.0        sale               S+3 avec jardin à chotrana 1    
1    380000.0        sale                         un s+2 à chotrana 1   
2    180000.0        sale                                S+3 en vente   
3    240000.0        sale  Charmant S+2 avec grande terrasse à vendre   
4    365000.0        sale            appartement s3 ennaser bon prix    
..        ...         ...                                         ...   
777  339300.0        sale                         A VENDRE S+2 SOUKRA   
778  380000.0        sale                    S+2 Sur Plan à La Soukra   
779  380000.0        sale                     S+3 en Vente à Ennasr 2   
780  550000.0        sale              S+3 à Vendre à Jardin d'Ennasr   
781  310000.0        sale                     S+3 en Vente à Ennasr 2   

        city         region  \
0     ariana     chotrana 1   
1     ariana     chotrana 1   
2    manouba  manouba ville   
3      tunis       gammarth   
4     ariana       ennasr 2   
..       ...            ...   
777   ariana         soukra   
778   ariana         soukra   
779   ariana       ennasr 2   
780   ariana         ennasr   
781   ariana       ennasr 2   

                                           description surface  bathrooms  \
0    Cet appartement est situé au rez-de-chaussée d...     141        3.0   
1    Cet appartement est situé à chotrana 1 dans un...     100        1.0   
2    Cet appartement se situe à Dyar ben Mahmoud Ag...     130        1.0   
3    À vendre : Appartement S+2 avec une grande ter...     120        1.0   
4    a vendre a ennaser un appartement s3 au 5éme é...     148        2.0   
..                                                 ...     ...        ...   
777  Notre agence palmarès immobilière met à la ven...     117        1.0   
778  ✅✅Palmarès immobilier met en vente un appartem...     115        1.0   
779  Cet appartement occupe le rez-de-chaussée d'un...     145        2.0   
780  Cet appartement occupe le troisième étage d’un...     176        3.0   
781  Cet appartement est situé au premier étage d'u...     119        2.0   

    rooms  
0       3  
1       2  
2       4  
3       1  
4      03  
..    ...  
777     3  
778     2  
779     3  
780     3  
781     3  

[648 rows x 9 columns]

##Identify rows with price 0.0, 1.0, 2.0, 2.5


In [131]:
# Capturing rows with price 0.0 or 1.0
filtered_df = df[df['price'].isin([0.0, 1.0 ,2.0,2.5])]

# Enumerating and extracting indexes
enumerated_indexes = list(enumerate(filtered_df.index))

print("\nEnumerated Indexes:")
print(enumerated_indexes)


Enumerated Indexes:
[(0, 16), (1, 27), (2, 28), (3, 36), (4, 100), (5, 139), (6, 154), (7, 164), (8, 166), (9, 177), (10, 183), (11, 195), (12, 213), (13, 214), (14, 252), (15, 275), (16, 289), (17, 335), (18, 351), (19, 375), (20, 376), (21, 437), (22, 444), (23, 459), (24, 460), (25, 516), (26, 542), (27, 558), (28, 567), (29, 574), (30, 585), (31, 588), (32, 589), (33, 590), (34, 612), (35, 616), (36, 622), (37, 631), (38, 632), (39, 648), (40, 684), (41, 704), (42, 710), (43, 732), (44, 743), (45, 756), (46, 766)]


# Manuel Report
Note: indexes are incremented by 2

**No price info**
- row index 18  (S+4 Lac ) -> no informations about the price

- row index 29  (APT kram s+2) -> no info about the price
- row index 30  (Complexe de loisir) -> not suitable for our case of studies
- row index 38  (APT chotrana 1) -> no info about the price
- row index 102 (S+2 Menzah 9C) -> no informations about the price
- row index 141 (APT Lafayette) -> no informations about the price
- row index 156 (4 appartements S+3) -> no info
- row index 166 (APT S+2 Ibnou Sina) -> no info
- row index 168 (APT Medina Jadida) -> no info
- row index 179 (APT S+3 Tunis) -> no info
- row index 185 (APTs Aouina) -> no info
- row index 197 (APT S+4 Aouina) -> no info
- row index 215 (Fouchana 4 APTs) -> négociable price LOL
- row index 216 (S+3 ghazela) -> no info
- row index 240-241 -> duplicate rows
- row index 254 (S+1 Jardin del Menzah)-> no info
- row index 277 (S+3 Chotrana) -> no info  
- row index 291 (APTs soukra aouina )-> no info
- row index 337 (APT Ennaser S+4) -> no info
- row index 353 (APT S+3 tunis) -> no info
- row index 377 (APT Ariana) -> no info
- row index 378 (APT S+3 Bardo) -> no info
- row index 461 (APT Menzah 1 S+4) -> no info
- row index 462 (APT chotrana1 S+3) -> no info
- row index 511 no clue
- row index 515 no clue
- row index 518 (APT mourouj 6) -> no info
- row index 536 no clue
- row index 544 (APT mourouj 5) -> no info
- row index 560 (APT medina Jadida )-> no clue
- row index 576 (APT medina Jadida )-> no clue
- row index 587 (APT manouba) -> no info
- row index 591 (APT hammam lif) -> no info
- row index 592 (APT ezzahra) -> no info
- row index 609 (APT borj Cedria ) -> no clue
- row index 614 (APTs + duplex ennassr2) -> no clue
- row index 624 (APT aouina) -> no info
- row index 634 (APT S+3 Menzah 6) -> no clue
- row index 734 (APT ain Zaghouan) -> no info
- row index 737 (APT soukra) -> no clue ??
- row index 745 (APT Mourouj 6 S+2) -> no clue ???
- row index 712 (APT chotrana)-> no info
---
**None ART type**
- row index 390 (Villa + autre bien Marsa) -> Drop row
- row index 439 (Villa) -> Drop row
- row index 446 (Villa) -> Drop row
- row index 752 (Villa) -> drop row
- row index 758 (Villa) -> drop row
- row index 776 (immeuble) -> drop row
- row index 459 (APT ariana) -> drop row
- row index 485 (Terrain Gammarth) -> drop row
- row index 640 (Villa ) -> drop row
- row index 678 (villa boumhal) -> drop row
- row index 618 (S+3 medina Jadida) ->  duplicate
- row index 633 (APT s+3 medina jedida)-> duplicate
- row index 667 (APT  manar 2 ) -> duplicate
- row index 686 (APT medina Jadida) -> duplicate

---
**For rent opportunities**
- row index 453 (APT soukra)-> For rent
- row index 655 (APT jardin de carthage) -> for rent
- row index 661 (APT Menzah 8) -> for rent
- row index 763 ( APT jardin del menzah ) -> for rent
---
**Calculation Needed**
- row index 474 (S+2 , S+3 Chotrana1 ) -> prix de mètre carré
- row index 555 (APTs + duplexes ain Zaghouan) -> 3500DT mètre carré
- row index 590 (APTS borj cedria) ->  prix de mètre carré
---

- row index 650 (APT Manouba) -> convert from 0.0 to 315.000DT
- **row index 699...702 -> price feature needs a revision**
- **row index 515 -> price feature needs a revision**
- row index 706 -> annonce de vente des appartements d'une résidence en construction
- row index 768 (APT mourouj 6 S+1) -> drop row
---
**Convertion needed to TND unit**
- row index 436 (borj louzir APT S+3) -> rectify price to 120.000TND
- row index 478 (APT manar 2) -> convert price to TND
- row index 491 (APT menzah9) -> convert price to TND
- row index 495-496-497-500-503 -> convert price from MDT to TND

- row index 632 (APT ain zaghouen)-> convert price to 210.000DT
- row index 638 (APT mourouj 5) -> convert to 118.000DT
- row index 649 (APT manar 2) ->convert to 420.000DT

- row index 654 (APT ariana) -> convert to 580.000DT
- row index 672 (APT soukra ) -> convert to 297.000DT
- row index 721 (APT Soukra)-> convert to 545.000DT
- row index 721 (APT Aouina)-> convert to 250.000DT
- row index 731 (APT mourouj 5)-> convert to 110.000DT
- row index 738 (APT la goulette) -> convert to 550.000DT
- row index 750 (APT ennasr) -> convert to 175.000DT



## Drop no price info annoucement rows and irrelevant price with surface feature equal 1

In [132]:
indexes_to_drop = [index for index in filtered_df.index if (index != 648 and index != 588) ]
indexes_to_drop.append(181)
indexes_to_drop.append(657)

# Dropping all indexes except index 648 (rows with 1.0, 0.0, 2.0, 2.5 prices)
clean_df = df.drop(indexes_to_drop, errors='ignore')

## Drop for rent announcement rows

In [134]:
# Dropping for rent rows
for_rent_to_drop = [451, 653, 659, 761,761,224, 293, 498, 513,295,322, 419, 509 , 735]
clean_df = clean_df.drop(for_rent_to_drop, errors='ignore')

## Drop None APT Type & non detected duplicates

In [135]:
# Dropping none APT rows and duplicate
indexes_to_drop = [437, 444, 750, 756, 774, 457, 483, 638, 676, 616, 631, 665, 684,291,388, 553,211]
clean_df = clean_df.drop(indexes_to_drop, errors='ignore')

## Convert unit to TND and correct some mistakes

In [136]:
# Filtering rows with price in the range of 100 to 999
price_filtered_df = clean_df[(clean_df['price'] >= 100) & (clean_df['price'] <= 999)]
print (len(price_filtered_df))
# Multiply the price by 100 for these rows
clean_df.loc[price_filtered_df.index, 'price'] *= 100
# Filtering rows with price over 5 Billion dinar
price_filtered_df2 = clean_df[clean_df['price'] >= 5000000]
print (len(price_filtered_df2))
# Divide the price by 100 for these rows
clean_df.loc[price_filtered_df2.index, 'price'] /= 100

26
4


In [137]:
display(price_filtered_df)

price transaction                                              title  \
18   120.0        sale                               APPARTEMENT A VENDRE   
122  180.0        sale                A Vendre appartement S+4 Montfleury   
212  260.0        sale                      S+3 madina jadida 3 à vendre    
333  540.0        sale          À vendre un appartement S +3 avec garage    
374  280.0        sale                            excellente opportunité    
392  194.0        sale                           S2 HS mourouj 6 a vendre   
413  350.0        sale     A vendre des appartements s1 direct promoteur    
418  270.0        sale                  APPARTEMENT S+3 à Vendre MENZAH 7   
434  120.0        sale                  Apt S+3 cité la santé Borj Louzir   
443  165.0        sale          A vendre idéal appartement s+3 Yasminatte   
493  450.0        sale                              apparemment s4 ennasr   
494  235.0        sale                             app s2 riadh andalous    
495  208.0        sale                      A vendre appartement menzah 7   
501  120.0        sale                                         شقة للبيع    
534  260.0        sale                                 Vente Appartement    
607  100.0        sale                                joli à Borj Cedria    
630  210.0        sale         A VENDRE UN APPARTEMENT S+1 A AIN ZAGHOUEN   
636  118.0        sale                   appartement a mourouj 5 a vendre   
647  111.0        sale                                       appartement    
652  580.0        sale                         Appartement à Ariana Ville   
670  297.0        sale                               A vendre S+1a soukra   
719  545.0        sale  VENTE APPARTEMENT STANDING S+3 A LA SOUKRA CÔT...   
727  250.0        sale                             À vendre s2 à l'Aouina   
729  110.0        sale                               Appart S+2 À vendre    
736  550.0        sale  kheireddine A Vendre appartement haut standing...   
748  175.0        sale                   A VENDRE S+1 POUR INVESTISSEMENT   

          city             region  \
18      ariana             ariana   
122      tunis          monfleury   
212  ben arous      medina jedida   
333     ariana             ennasr   
374  ben arous            ezzahra   
392  ben arous          mourouj 6   
413      tunis  ain zaghouan nord   
418     ariana           menzah 7   
434     ariana        borj louzir   
443  ben arous          ben arous   
493     ariana           ennasr 2   
494     ariana     riadh andalous   
495     ariana           menzah 7   
501    manouba            manouba   
534      tunis              tunis   
607  ben arous        borj cedria   
630      tunis       ain zaghouan   
636  ben arous          mourouj 5   
647      tunis            manar 2   
652     ariana       ariana ville   
670     ariana             soukra   
719     ariana             soukra   
727      tunis             aouina   
729  ben arous          mourouj 5   
736      tunis        la goulette   
748     ariana             ennasr   

                                           description surface  bathrooms  \
18   A vendre appartement S+1 à citer takacim borj ...      54        1.0   
122  A vendre Appartement S+4 sur Rue du Sahel près...      97        1.0   
212  📌 Offre de vente Appartement HS situé à #Madin...     122        2.0   
333  L'agence modern immo met à la vente un apparte...     202        2.0   
374  je mets en vente un appartement s+3 haut stand...     132        2.0   
392  💥OFFRE DE VENTE APPARTEMENT HS 📍#Mourouj6 🎈App...      78        1.0   
413  Mr casa vous propose pour la vente des apparte...      60        1.0   
418  Notre agence met en vente un appartement s+3 a...     106        2.0   
434  VAA 006 A vendre un Appartement S+3 à cité la ...      78        1.0   
443  A vendre appartement s+3 2èm etage yasminat se...     115        1.0   
493  a vendre à ennasr 2 avenue hedi nouira apparte...     135        2.0   
4

In [138]:
display(price_filtered_df2)

price transaction                                   title     city  \
377  150000000.0        sale          A vendre Appartement à manouba    tunis   
442  157000000.0        sale                               شقة للبيع  manouba   
476  280000000.0        sale             appartement vente a Manar2     tunis   
489  320000000.0        sale  appartement vente au coeur de Menzah9     tunis   

       region                                        description surface  \
377     bardo  je mets en vente un appartement situé au cité ...     200   
442    denden  للبيع شقة بالطابق الأول متكونة من غرفتين و قاع...      90   
476   manar 2  La Rosa immobilière met en vente un appartemen...     110   
489  menzah 9  La Rosa immobilière met en vente un appartemen...     131   

     bathrooms rooms  
377        1.0     3  
442        2.0     3  
476        1.0     3  
489        1.0     4

In [139]:
# Set the correct price
if 647 in clean_df.index:
    clean_df.at[647, 'price'] = 420000
if 648 in clean_df.index:
    clean_df.at[648, 'price'] = 315000
if 588 in clean_df.index:
    clean_df.at[588, 'price'] = 2300*136

In [140]:
# Find rows where the price is not a multiple of 10^5
not_multiple_of_10_5_df = clean_df[(clean_df['price'] // 10**5 == 0) & (clean_df['price'] // 10**4 != 0)   ]

display(not_multiple_of_10_5_df)

price transaction                                              title  \
18   12000.0        sale                               APPARTEMENT A VENDRE   
122  18000.0        sale                A Vendre appartement S+4 Montfleury   
207  95000.0        sale       Bab dzira A vendre un joli appartement S+2\n   
212  26000.0        sale                      S+3 madina jadida 3 à vendre    
302  95000.0        sale                              appartement a vendre    
333  54000.0        sale          À vendre un appartement S +3 avec garage    
374  28000.0        sale                            excellente opportunité    
392  19400.0        sale                           S2 HS mourouj 6 a vendre   
413  35000.0        sale     A vendre des appartements s1 direct promoteur    
418  27000.0        sale                  APPARTEMENT S+3 à Vendre MENZAH 7   
434  12000.0        sale                  Apt S+3 cité la santé Borj Louzir   
443  16500.0        sale          A vendre idéal appartement s+3 Yasminatte   
493  45000.0        sale                              apparemment s4 ennasr   
494  23500.0        sale                             app s2 riadh andalous    
495  20800.0        sale                      A vendre appartement menzah 7   
501  12000.0        sale                                         شقة للبيع    
534  26000.0        sale                                 Vente Appartement    
607  10000.0        sale                                joli à Borj Cedria    
630  21000.0        sale         A VENDRE UN APPARTEMENT S+1 A AIN ZAGHOUEN   
636  11800.0        sale                   appartement a mourouj 5 a vendre   
652  58000.0        sale                         Appartement à Ariana Ville   
670  29700.0        sale                               A vendre S+1a soukra   
692  78000.0        sale                       Appartement S+2 A Ksar Said    
694  99000.0        sale                Appartement S+3 A Ksar Said 2 Bardo   
719  54500.0        sale  VENTE APPARTEMENT STANDING S+3 A LA SOUKRA CÔT...   
727  25000.0        sale                             À vendre s2 à l'Aouina   
729  11000.0        sale                               Appart S+2 À vendre    
736  55000.0        sale  kheireddine A Vendre appartement haut standing...   
748  17500.0        sale                   A VENDRE S+1 POUR INVESTISSEMENT   

          city             region  \
18      ariana             ariana   
122      tunis          monfleury   
207      tunis              tunis   
212  ben arous      medina jedida   
302      tunis              tunis   
333     ariana             ennasr   
374  ben arous            ezzahra   
392  ben arous          mourouj 6   
413      tunis  ain zaghouan nord   
418     ariana           menzah 7   
434     ariana        borj louzir   
443  ben arous          ben arous   
493     ariana           ennasr 2   
494     ariana     riadh andalous   
495     ariana           menzah 7   
501    manouba            manouba   
534      tunis              tunis   
607  ben arous        borj cedria   
630      tunis       ain zaghouan   
636  ben arous          mourouj 5   
652     ariana       ariana ville   
670     ariana             soukra   
692      tunis          ksar said   
694      tunis          ksar said   
719     ariana             soukra   
727      tunis             aouina   
729  ben arous          mourouj 5   
736      tunis        la goulette   
748     ariana             ennasr   

                                           description surface  bathrooms  \
18   A vendre appartement S+1 à citer takacim borj ...      54        1.0   
122  A vendre Appartement S+4 sur Rue du Sahel près...      97        1.0   
207  A vendre joli appartement S+2 ,se situe à bab ...      51        1.0   
212  📌 Offre de vente Appartement HS situé à #Madin...     122        2.0   
302  llbi3 appartement s+2 m9abel beb el falla a cô...      51        1.0   
333  L'agence modern immo met à la vente un apparte...     202 

In [141]:
# Multiply the price by 10 the price except for indexes 207, 302, 692, and 694
indexes_to_exclude = [207, 302, 692, 694]
indexes_to_update = not_multiple_of_10_5_df.index.difference(indexes_to_exclude)
clean_df.loc[indexes_to_update, 'price'] *= 10

In [142]:
# Find rows where the price is not a multiple of 10^4
not_multiple_of_10_4_df = clean_df[(clean_df['price'] // 10**4 == 0)  ]

display(not_multiple_of_10_4_df)

price transaction                                              title  \
149  2850.0        sale                À vendre s2 et S3 neuf à la soukra    
258  1050.0        sale  un appartement S+1 à la nouvelle soukra avec p...   
472  3100.0        sale     À vendre s2 et S3 neuf à la soukra chotrana 1    

       city      region                                        description  \
149  ariana      soukra  #S+2/s+3 La Soukra#vente(2850dtm2 TTC) On vous...   
258   tunis      aouina  Agence capital immo vous propose a la location...   
472  ariana  chotrana 1  #S+2/s+3 #Soukra #Chotrana1 #vente On vous pro...   

    surface  bathrooms rooms  
149     124        2.0     2  
258      60        1.0     1  
472     120        2.0     2

In [143]:
# Update the price feature by multiplying the current value with the surface feature value for specified rows
for index in not_multiple_of_10_4_df.index:
    if index in df.index:
        clean_df.at[index, 'price'] *= float(clean_df.at[index, 'surface'])

In [144]:
# Verify the existence of rows with indexes 472, 553, and 588 and display them
indexes_to_verify = [472, 553, 588]
existing_rows = df.loc[df.index.intersection(indexes_to_verify)]

print("Existing rows with specified indexes:")
display(existing_rows)

Existing rows with specified indexes:


price transaction                                              title  \
472  3100.0        sale     À vendre s2 et S3 neuf à la soukra chotrana 1    
553  3500.0        sale  Des appartements S0, S+1, S+2 ou S+3 & Duplexe...   
588     1.0        sale            A vendre des appartement à Borej cedria   

          city        region  \
472     ariana    chotrana 1   
553      tunis  ain zaghouan   
588  ben arous   borj cedria   

                                           description surface  bathrooms  \
472  #S+2/s+3 #Soukra #Chotrana1 #vente On vous pro...     120        2.0   
553  Situé à Ain Zaghouan, ce projet résidentiel es...  48.148        1.0   
588  A vendre à Borej cedria prés de hôtel “SOLIMAR...     136        1.0   

    rooms  
472     2  
553     2  
588     3

## Reduid index and save the Cleaned dataset

In [145]:
display (clean_df)

price transaction                                       title  \
0    490000.0        sale               S+3 avec jardin à chotrana 1    
1    380000.0        sale                         un s+2 à chotrana 1   
2    180000.0        sale                                S+3 en vente   
3    240000.0        sale  Charmant S+2 avec grande terrasse à vendre   
4    365000.0        sale            appartement s3 ennaser bon prix    
..        ...         ...                                         ...   
777  339300.0        sale                         A VENDRE S+2 SOUKRA   
778  380000.0        sale                    S+2 Sur Plan à La Soukra   
779  380000.0        sale                     S+3 en Vente à Ennasr 2   
780  550000.0        sale              S+3 à Vendre à Jardin d'Ennasr   
781  310000.0        sale                     S+3 en Vente à Ennasr 2   

        city         region  \
0     ariana     chotrana 1   
1     ariana     chotrana 1   
2    manouba  manouba ville   
3      tunis       gammarth   
4     ariana       ennasr 2   
..       ...            ...   
777   ariana         soukra   
778   ariana         soukra   
779   ariana       ennasr 2   
780   ariana         ennasr   
781   ariana       ennasr 2   

                                           description surface  bathrooms  \
0    Cet appartement est situé au rez-de-chaussée d...     141        3.0   
1    Cet appartement est situé à chotrana 1 dans un...     100        1.0   
2    Cet appartement se situe à Dyar ben Mahmoud Ag...     130        1.0   
3    À vendre : Appartement S+2 avec une grande ter...     120        1.0   
4    a vendre a ennaser un appartement s3 au 5éme é...     148        2.0   
..                                                 ...     ...        ...   
777  Notre agence palmarès immobilière met à la ven...     117        1.0   
778  ✅✅Palmarès immobilier met en vente un appartem...     115        1.0   
779  Cet appartement occupe le rez-de-chaussée d'un...     145        2.0   
780  Cet appartement occupe le troisième étage d’un...     176        3.0   
781  Cet appartement est situé au premier étage d'u...     119        2.0   

    rooms  
0       3  
1       2  
2       4  
3       1  
4      03  
..    ...  
777     3  
778     2  
779     3  
780     3  
781     3  

[579 rows x 9 columns]

In [146]:
# Rebuild the index of clean_df
clean_df.reset_index(drop=True, inplace=True)

# Save the DataFrame to a CSV file with the name tayara_tn_clean
clean_df.to_csv('tayara_tn_clean.csv', index=False)